PV node: feature #185

scud-soptim opened the pull request:
Fixes #185

### Changes proposed in this PR include:

> This PR introduces a dedicated `voltage_regulator` component to enable voltage-controlled (PV) nodes in the power flow model. The component expects the attribute `u_ref` and enforces the node voltage magnitude to be regulated to this reference value.
>
> To support this behavior, the mathematical formulation and corresponding Jacobian matrix extensions for PV nodes were implemented. In addition, new validation tests and datasets were created to verify correct PV-node behavior across relevant scenarios. The user-facing documentation for the `voltage_regulator` was updated accordingly.

* Added new component `voltage_regulator` (requires attribute `u_ref`) to introduce PV-node behavior.
* Implemented PV-node voltage regulation constraint and integrated it into **newton-raphson** solver workflow.
* Extended the Jacobian matrix formulation to support PV-node equations consistently.
* Added dedicated validation tests and new test datasets to cover PV-node behavior.
* Updated documentation for `voltage_regulator` to reflect usage, parameters, and expected behavior.

### Fixes
> Accounting for shunt elements in the calculation of source results.

### Follow-Up Tasks
* Q-limit handling with PV→PQ switching and water-filling–based reactive power distribution among generators.

frie-soptim committed:
#185 initial PV Node implementation

Signed-off-by: Eduard Fried <eduard.fried@soptim.de>

# Conflicts:
#	power_grid_model_c/power_grid_model/include/power_grid_model/math_solver/common_solver_functions.hpp

frie-soptim committed:
#185 adjust result calculation for pv generators and source

- set pq load gens from input
- for pv gens set P (from input) and distribute Q (from bus injection)
- push rest of P and Q to source
  - if both PV gen and source conntected to a bus, source will only get P

Signed-off-by: Eduard Fried <eduard.fried@soptim.de>

frie-soptim committed:
#185 add asymmetric case for distribution of q on pv generators

Signed-off-by: Eduard Fried <eduard.fried@soptim.de>

frie-soptim committed:
#185 improve calculation of jacobian matrix, distribute bus injections onto pv generators

Signed-off-by: Eduard Fried <eduard.fried@soptim.de>

# Conflicts:
#	power_grid_model_c/power_grid_model/include/power_grid_model/math_solver/common_solver_functions.hpp

frie-soptim committed:
#185 handle disconnected pv generators by changing their type to const_pq

Signed-off-by: Eduard Fried <eduard.fried@soptim.de>

frie-soptim committed:
#185 simplify delta q calculation for pv bus

- set delta q to zero at pv bus for all load/generators, not just pv generators

Signed-off-by: Eduard Fried <eduard.fried@soptim.de>

frie-soptim committed:
#185 extend update object for voltage_regulation

- add u_ref
- add q_min and q_max, as they may be dependent on the specified active power value of the generator

Signed-off-by: Eduard Fried <eduard.fried@soptim.de>

frie-soptim committed:
#185 add python tests

Signed-off-by: Eduard Fried <eduard.fried@soptim.de>

frie-soptim committed:
- add new component voltage_regulator and the necessary boilerpate
- remove previous implementation with the const_pv value

Signed-off-by: Eduard Fried <eduard.fried@soptim.de>

frie-soptim committed:
#185 add documentation for the voltage_regulator

Signed-off-by: Eduard Fried <eduard.fried@soptim.de>

frie-soptim committed:
#185 added test networks for pv node tests

Signed-off-by: Eduard Fried <eduard.fried@soptim.de>

frie-soptim committed:
fix windows build

Signed-off-by: Eduard Fried <eduard.fried@soptim.de>

scud-soptim committed:
Merge branch 'main' into feature/#185PV-Node

scud-soptim commented:
We do not have sufficient permissions to set the label on this repository.

mgovers commented on docs/user_manual/components.md:
```diff
@@ -1245,3 +1245,74 @@ node_1 --- transformer_4 --- node_2 --- line_5 --- node_3
+| name              | data type         | unit                       | description                                                          |
+| ----------------- | ----------------- | -------------------------- | -------------------------------------------------------------------- |
+| `q`               | `RealValueOutput` | volt-ampere-reactive (var) | reactive power provided by the voltage regulator                     |
+| `limit_violated`  | `int8_t`          | -                          | reactive power limit violation indicator (not yet fully implemented) |
```
unrelated to this PR but maybe we should add such a flag to the tap regulator as well. Relates to #1215

mgovers commented on docs/user_manual/components.md:
```diff
@@ -1245,3 +1245,74 @@ node_1 --- transformer_4 --- node_2 --- line_5 --- node_3
+   \end{eqnarray}
+$$
+
+In this case, `limit_violated` will indicate which limit was exceeded, and the actual voltage at the node may differ from `u_ref`.
```
since this is not fully implemented yet, do we want to explicitly mention the current behavior with a big warning note that this is temporary behavior?

scud-soptim replied:
a warning-note!

mgovers commented on power_grid_model_c/power_grid_model/include/power_grid_model/calculation_parameters.hpp:
```diff
@@ -158,6 +169,9 @@ struct MathModelTopology {
     DenseGroupedIdxVector tap_regulators_per_branch;
 
+    // TODO: #185 There should be at most one voltage regulator per load_gen. Is there a simpler way to represent this?
+    DenseGroupedIdxVector voltage_regulators_per_load_gen;
```
The current way is OK.

The way that `DenseGroupedIdxVector` works is that it already uses an efficient representation, so both memory-wise and when looping over the data, it optimizes for space.

In particular, rather than storing something along the lines of `{load_gen_1: [regulator_1], load_gen_2: [], load_gen_3: [], load_gen_4: [regulator_2]}`, it stores it like `{"regulator_1": [load_gen_1], "regulator_2: [load_gen_4]}`

C++23 `std::optional` should be able to optimize this code even further, but we haven't gotten to that yet.

P.s.: we welcome better naming conventions!

scud-soptim replied:
we remove the TODO Statement.

mgovers commented on power_grid_model_c/power_grid_model/include/power_grid_model/component/voltage_regulator.hpp:
```diff
@@ -0,0 +1,118 @@
+    double injection_direction() const { return injection_direction_; }
+    double u_ref() const { return u_ref_; }
+    double q_min() const { return q_min_; }
+    double q_max() const { return q_max_; }
```
please make `constexpr`

mgovers commented on power_grid_model_c/power_grid_model/include/power_grid_model/main_core/output.hpp:
```diff
@@ -451,6 +458,39 @@ output_result(Component const& transformer_tap_regulator, MainModelState<Compone
+        return state.topo_comp_coup->load_gen[obj_seq];
+    }();
+    if (gen_math_id.group != -1) {
+        // is voltage regulator always in same group as the generator it regulates?
```
yes because the group represents the connected component (in the graph-theoretical sense) it belongs to. That is, if you have 2 distinct subgrids that are not connected to eachother, then you will have 2 topological groups, and therefore also 2 math solvers (one for each subgrid).

As a special case, if a subgrid does not contain a source, the subgrid is considered disconnected, will not have an associated math solver, and therefore, the `group` for all components in that subgrid will be `-1`

if the groups are not equal, there's a bug in the code concerning topology

mgovers reviewed (commented):
Hi @scud-soptim, Nice to see this coming together the way it is.

I've gone through most of the interface and boilerplate code. I'll continue with the review of the actual solver and input/output conversions on Monday, but here is already some input.

mgovers commented on power_grid_model_c/power_grid_model/include/power_grid_model/math_solver/common_solver_functions.hpp:
```diff
@@ -125,14 +127,19 @@ inline void calculate_multiple_source_result(IdxRange const& sources, YBus<asymm
 inline void calculate_source_result(IdxRange const& sources, Idx const& bus_number, YBus<sym> const& y_bus,
                                     PowerFlowInput<sym> const& input, SolverOutput<sym>& output,
-                                    IdxRange const& load_gens) {
+                                    IdxRange const& load_gens, IdxRange const& shunts) {
```
EDIT: i now see that the description contains

> Accounting for shunt elements in the calculation of source results.

Would it be possible to put that in a separate PR including a validation case? This looks like a bug that deserves first-class treatment.

figueroa1395 replied:
> [@scud-soptim](https://github.com/scud-soptim) Shunt gets treated as part of ybus matrix so its contribution should already be included via output.bus_injection.


That is true, and that is the total current in the bus. Here the individual shunt contribution needs to be calculated in order to isolate the source current contribution. So I don't think this is redundant, is it @nitbharambe?

scud-soptim commented on power_grid_model_c/power_grid_model/include/power_grid_model/calculation_parameters.hpp:
```diff
@@ -79,6 +79,17 @@ template <symmetry_tag sym_type> struct ApplianceShortCircuitSolverOutput {
+    RealValue<sym> q{};
+
+    // provide generator info, as the regulator component has no other access to it
+    ID generator_id{};
```
We find to out ....

frie-soptim replied:
I use the generator_id in the `output_result` result function.
```
    if (load_gen_math_id.group != -1) {
        // is voltage regulator always in same group as the generator it regulates?
        for (auto const& vr_output: math_output.solver_output[load_gen_math_id.group].voltage_regulator) {
            if (vr_output.generator_id == voltage_regulator.regulated_object()) {
                return voltage_regulator.template get_output<sym>(vr_output);
            }
        }
    }
```
Unfortunately, i an not familiar enough with the internal topology structure to easily work without this id.

figueroa1395 commented on power_grid_model_c/power_grid_model/include/power_grid_model/main_core/topology.hpp:
```diff
@@ -169,11 +169,16 @@ constexpr void register_topology_components(ComponentContainer const& components
 
+    apply_registration<Component>(components, comp_topo.regulator_type,
+                                  [](Regulator const& regulator) { return regulator.math_model_type(); });
+
```
Why is this now needed? This change also affects `TransformerTapRegulator`.

frie-soptim replied:
not needed, removed

figueroa1395 commented on power_grid_model_c/power_grid_model/include/power_grid_model/math_solver/iterative_pf_solver.hpp:
```diff
@@ -90,13 +91,17 @@ template <symmetry_tag sym, typename DerivedSolver> class IterativePFSolver {
     std::shared_ptr<DoubleVector const> phase_shift_;
     std::shared_ptr<SparseGroupedIdxVector const> load_gens_per_bus_;
     std::shared_ptr<DenseGroupedIdxVector const> sources_per_bus_;
+    std::shared_ptr<DenseGroupedIdxVector const> shunts_per_bus_;
```
Given https://github.com/PowerGridModel/power-grid-model/pull/1220/changes#r2617541914, I'll post this here as a reminder to remove `shunts_per_bus_` here and in the surroundings.

figueroa1395 reviewed (commented):
I'm done reviewing the whole PR. Some additional side remarks:
- When `q` limits are fully implemented, don't forget the "cooldown" factor to avoid back and forth between PV and PQ nodes. In addition, don't forget to add `q` limits to the tests as well.
- Don't forget to check whether power flow with automatic tap regulators work seamlessly with PV nodes now. Probably some unit + validation test (if it works) are a good idea.
- It would be nice to have some unit tests on combinations of having multiple regulators with the same `u_ref` in the same node, different `u_ref`, with both `a/sym_gen` `a/sym_load_gen`, etc., in addition to the validation tests. Moreover, unit tests of other methods besides `newton_raphson` raising (was this behavior agreed upon already?) when `voltage_regulator`s are present are also welcome (I believe the last can be easily done in the `params.json` by including the other calculation methods and marking them as expected to fail).

nitbharambe commented on power_grid_model_c/power_grid_model/include/power_grid_model/component/voltage_regulator.hpp:
```diff
@@ -0,0 +1,118 @@
+            .status = static_cast<IntS>(status()),
+            .u_ref = {u_ref_, 0.0},
+            .q_min = RealValue<sym>{q_min_}, // TODO: #185 divide by 3 for asymmetric case?
+            .q_max = RealValue<sym>{q_max_}, // TODO: #185 divide by base_power?
```
Yes, this attribute is in p.u. internally so both can be divided by `base_power_3p`. 

These variables here can denote total limit, ie. sum of all phases power.
Further division by 3 would depend on whether loadgen is symmetric or asymmetric type. Let's handle that at either param calculation or solver side where we have information on the regulated object.

frie-soptim committed:
#185 changes from code review

Signed-off-by: Eduard Fried <eduard.fried@soptim.de>

frie-soptim committed:
#185 declare voltage regulation as an experimental feature for now

- until PV->PQ conversion is implemented

Signed-off-by: Eduard Fried <eduard.fried@soptim.de>

frie-soptim committed:
#185 add more unit/validation tests

Signed-off-by: Eduard Fried <eduard.fried@soptim.de>

frie-soptim committed:
#185 update documentation

Signed-off-by: Eduard Fried <eduard.fried@soptim.de>

frie-soptim committed:
fix code quality error

Signed-off-by: Eduard Fried <eduard.fried@soptim.de>

frie-soptim committed:
#185 move voltage regulator code into newton_raphson_pf_solver

- calculate calculate_pf_result() function need access to voltage regulators
  -> read them directly from y_bus, instead of passing them via function arguments
- remove references to voltage regulators (and shunts) that were passed on to the result() function in other solvers

Signed-off-by: Eduard Fried <eduard.fried@soptim.de>

frie-soptim committed:
update tests

Signed-off-by: Eduard Fried <eduard.fried@soptim.de>

scud-soptim commented on power_grid_model_c/power_grid_model/include/power_grid_model/math_solver/newton_raphson_pf_solver.hpp:
```diff
@@ -357,6 +396,40 @@ class NewtonRaphsonPFSolver : public IterativePFSolver<sym_type, NewtonRaphsonPF
+        //   - all off-diagonal voltage derivatives (L block) are zero,
+        //   - all θ-derivatives (M block) are zero,
+        //   - only the diagonal L entry is set to V.
+        // This removes the Q-equation and enforces the PV voltage constraint in the NR step.
```
You are right that, numerically, setting the Jacobian diagonal to 1.0 would lead to equivalent behavior in this formulation: since the PV row replaces the Q equation and we start from U = U_ref, the Newton update keeps ΔU = 0 as long as the bus remains PV.

The reason for explicitly setting the PV row to L = v (and zeroing the rest of the row) is mainly semantic and robustness-related: it makes it explicit in the Jacobian that this equation is an identity constraint enforcing V = Vset, not a physical power balance. This becomes important for readability, debugging, and for later extensions (e.g. PV→PQ switching, Q-limit handling), where it is useful to clearly distinguish “ignored equations” from active ones.

So while diag = 1.0 is mathematically sufficient, the current form is intentionally descriptive rather than minimal.

frie-soptim commented on power_grid_model_c/power_grid_model/include/power_grid_model/component/voltage_regulator.hpp:
```diff
@@ -0,0 +1,118 @@
+            q_max_ = new_q_max;
+            changed = true;
+        }
+        return changed;
```
`changed` not needed, removed

frie-soptim commented on power_grid_model_c/power_grid_model/include/power_grid_model/math_solver/newton_raphson_pf_solver.hpp:
```diff
@@ -234,13 +237,47 @@ class NewtonRaphsonPFSolver : public IterativePFSolver<sym_type, NewtonRaphsonPF
+        }
+
+        // expect that one load/gen is only regulated by one regulator and if multiple loads/gens with regulators
+        // are connected to the same bus, they all have the same u_ref
```
done

frie-soptim committed:
fix code quality error

Signed-off-by: Eduard Fried <eduard.fried@soptim.de>

scud-soptim committed:
Merge branch 'main' into feature/#185PV-Node

scud-soptim commented:
"Goedemorgen". I have a question about resolving PR request comments. How do you handle that? Who should resolve a dialogue?

frie-soptim committed:
fix more code quality errors

- add line breaks in documentation

Signed-off-by: Eduard Fried <eduard.fried@soptim.de>

scud-soptim committed:
Merge branch 'main' into feature/#185PV-Node

frie-soptim committed:
#185 fix result calculation

- revert calculation of source values to previous state
- shunts are not included in the bus injection, only load and generation are

Signed-off-by: Eduard Fried <eduard.fried@soptim.de>

figueroa1395 commented on power_grid_model_c/power_grid_model/include/power_grid_model/main_core/calculation_input_preparation.hpp:
```diff
@@ -140,18 +153,28 @@ void prepare_input_status(main_model_state_c auto const& state, std::vector<Idx2
+        state, state.topo_comp_coup->voltage_regulator, pf_input,
+        [&state](Idx i) { return state.comp_topo->regulated_object_type[i] == ComponentType::generic_load_gen; });
+
+    prepare_input_status<sym, &PowerFlowInput<sym>::load_gen_status, GenericLoadGen>(
```
Results of offline discussion: This can stay as is for now. If we decide to refactor at some point, we will pick it up as a follow up. Reason: It is not clear if going via topology instead is a good idea performance wise, so we aren't sure if that is worth it for now.

figueroa1395 commented on power_grid_model_c/power_grid_model/include/power_grid_model/component/voltage_regulator.hpp:
```diff
@@ -0,0 +1,113 @@
+
+    bool is_energized(bool is_connected_to_source = true) const {
+        return is_connected_to_source && status();
+    }
```
Can you make this function `constexpr`?

frie-soptim replied:
done

figueroa1395 reviewed (commented):
>Scope of this PR
This PR is intentionally limited to introducing PV nodes and their basic handling. Q-limit enforcement (including PV→PQ switching, cooldown, and hysteresis to avoid oscillations) is considered a separate concern and should, in my view, be handled independently to keep responsibilities and changes clearly separated.
>
>Q-limits / cooldown / hysteresis
I fully agree that cooldown / hysteresis is essential once Q-limits are enforced. To keep concerns cleanly separated, it would be great to track this as a dedicated issue and follow-up PR. If you are open to that, you could create a separate issue for Q-limit handling and assign it to us.

The issue has been created and assigned to you. See #1236 .

Let's indeed keep the scope of this PR as small as possible and do incremental improvements.

>Interaction between tap changers and voltage regulators
Regarding the interaction of automatic tap changers and voltage regulators: having both acting on the same bus is, from a modeling perspective, not well-defined. A tap changer can modify the bus voltage via the transformer, while a voltage regulator enforces a fixed voltage setpoint at that same bus, which can lead to conflicting control objectives.
>
>Rather than only testing this combination, it would therefore be preferable to introduce a validation check that detects and rejects configurations where a tap changer and a voltage regulator operate on the same node.

You are right: let's not deal with this issue for now. Can you `throw` a hard error/crash in the core when automatic tap regulation is used in combination with voltage regulators? In addition, can you also add this check to the validation function.

On the other hand, I've resolved several comments and made a few additional minor ones.

frie-soptim committed:
#185 throw if voltage regulators not valid

- if there are voltage regulators and transformer tap regulators in one model
- if there are two voltage regulators that regulate the same node, but have different u_ref values

Signed-off-by: Eduard Fried <eduard.fried@soptim.de>

frie-soptim commented on power_grid_model_c/power_grid_model/include/power_grid_model/math_solver/newton_raphson_pf_solver.hpp:
```diff
@@ -296,6 +302,46 @@ class NewtonRaphsonPFSolver : public IterativePFSolver<sym_type, NewtonRaphsonPF
+        std::map<Idx, Idx> loadgen_to_regulator = prepare_mapping_of_active_regulators(input);
+
+        // expect that one load/gen is only regulated by one regulator and if multiple loads/gens with regulators
+        // are connected to the same bus, they all have the same u_ref
```
done. 
Added a function in main_model_impl.hpp, not sure if there is a better place for this.
Two regulators on an appliance already threw a DuplicativelyRegulatedObject.

figueroa1395 replied:
>Added a function in main_model_impl.hpp, not sure if there is a better place for this.

I'll resolve this comment and give further feedback at the implementation.

>Two regulators on an appliance already threw a DuplicativelyRegulatedObject.

You are correct.

frie-soptim commented on src/power_grid_model/validation/_validation.py:
```diff
@@ -342,10 +342,14 @@ def validate_required_values(  # noqa: PLR0915
+    required[ComponentType.voltage_regulator] = required["regulator"].copy()
+    if calculation_type is None or calculation_type == CalculationType.power_flow:
+        required[ComponentType.voltage_regulator] += ["u_ref"]
+
```
done

frie-soptim committed:
add missing include

Signed-off-by: Eduard Fried <eduard.fried@soptim.de>

frie-soptim committed:
fix compiler errors

Signed-off-by: Eduard Fried <eduard.fried@soptim.de>

frie-soptim committed:
fix formatting

Signed-off-by: Eduard Fried <eduard.fried@soptim.de>

figueroa1395 commented on power_grid_model_c/power_grid_model/include/power_grid_model/component/voltage_regulator.hpp:
```diff
@@ -0,0 +1,106 @@
+    VoltageRegulatorOutput<sym> get_output(VoltageRegulatorSolverOutput<sym> const& solver_output) const {
+        VoltageRegulatorOutput<sym> output{};
+        static_cast<BaseOutput&>(output) = base_output(is_energized(true) && solver_output.generator_status != 0);
+        output.limit_violated = solver_output.limit_violated;
```
This is a question for you as users.

Should we make `limit_violated` an `enum` instead? With values such as `PV_operation`, `PQ_operation` or something similar (open to suggestions)? Should we add if the lower or upper bound was violated? Or just keep as is? What is your use case?

frie-soptim replied:
Once PV->PQ conversion is implemented, it was intended to use an enum value to indicate whether the q_min or q_max limit were violated. As this cannot happen in the current state of the implementation, the values is always 0.

frie-soptim committed:
fix formatting again

- now with the correct clang-format version

Signed-off-by: Eduard Fried <eduard.fried@soptim.de>

figueroa1395 reviewed (commented):
Some additional remarks:
- We'll remove the feature flag after the holidays. That way both parties can implement + test more easily and we start adding immediate value for users in general.
- Can you provide validation tests in which multiple `load_gen`s of different types (`const_impedance`, `const_current`, and `const_power`) are present?
- Can you also provide validation tests for a simple grid in which a regulated and a non-regulated `sym_load` are at the same node?

figueroa1395 committed:
implement check_voltage_regulator_validity using topology

Signed-off-by: Santiago Figueroa Manrique <figueroa1395@gmail.com>

figueroa1395 committed:
refactor set_u_ref_and_bus_types without additional maps

Signed-off-by: Santiago Figueroa Manrique <figueroa1395@gmail.com>

scud-soptim committed:
Merge pull request #1 from PowerGridModel/feature/pv-node-topology-refactor

PV node support: Small refactor - Use internal topology

nitbharambe commented:
> The voltage-regulator–based solution is more consistent with the CIM/CGMES model, since a synchronous machine can operate either as a generator or as a motor (SSH profile). Consequently, this property must be updateable at runtime. This requirement would not be adequately supported by an enum-based approach.

Ok that makes sense. And based on the requirements, lets skip the addition of enum. @figueroa1395 you can proceed with https://github.com/scud-soptim/power-grid-model/pull/1

frie-soptim committed:
#185 fix bus injection distribution onto regulating loads/generators, update tests

- ignore specified p,q of voltage regulating load/generations during distribution
- adjust handling of enabled sources and voltage regulators at the same node, throw only when both are enabled
- add new 6 node test dataset with various combinations of regulated loads/generations
- add test datasets for unsupported cases
  - voltage regulators and transformer tap regulators in model
  - enabled source and enabled voltage regulator on same nome

Signed-off-by: Eduard Fried <eduard.fried@soptim.de>

frie-soptim committed:
#185 remove property q from voltage_regulator output

- the contribution of the regulator is directly representen in the output of the load/generator
- voltage_regulator.q was redundant

Signed-off-by: Eduard Fried <eduard.fried@soptim.de>

mgovers commented on power_grid_model_c/power_grid_model/include/power_grid_model/math_solver/newton_raphson_pf_solver.hpp:
```diff
@@ -357,6 +391,39 @@ class NewtonRaphsonPFSolver : public IterativePFSolver<sym_type, NewtonRaphsonPF
+        // For PV buses, the Q-equation is replaced by the algebraic constraint
+        //     rPV = V - Vset = 0.
+        // Since the state vector uses relative voltage increments ΔV_rel = ΔV / V,
+        // the derivative ∂rPV/∂ΔV_rel equals the current voltage magnitude V.
```
here and elsewhere: please make sure that there are only ASCII characters in source files. This means that e.g. ∂ and θ are not allowed.

frie-soptim replied:
removed non-ascii characters

frie-soptim committed:
removed non-ascii characters

Signed-off-by: Eduard Fried <eduard.fried@soptim.de>

figueroa1395 committed:
Merge branch 'main' into feature/#185PV-Node

figueroa1395 committed:
fix check_voltage_regulator_validity

Signed-off-by: Santiago Figueroa Manrique <figueroa1395@gmail.com>

figueroa1395 commented on power_grid_model_c/power_grid_model/include/power_grid_model/calculation_preparation.hpp:
```diff
@@ -131,6 +133,80 @@ void rebuild_topology(typename ModelType::MainModelState& state, SolverPreparati
+                    // TODO(figueroa1395): test this
+                    if (u_refs_at_bus.size() > 1) {
+                        throw ConflictingVoltageRegulatorURef{
+                            std::format("{}, {}", first_voltage_regulator_id, voltage_regulator_at_node_id)};
```
Hello @frie-soptim. I found the issue and I fixed it in b7a7c67. My previous implementation of the check was flawed. Now the tests should pass accordingly. Please have a look.

I will later implement some additional unit tests for such functions, but that's on us and not for you to worry about.

figueroa1395 commented on power_grid_model_c/power_grid_model/include/power_grid_model/main_model_impl.hpp:
```diff
@@ -486,7 +487,87 @@ class MainModelImpl {
+                if (u_refs.size() > 1) {
+                    u_refs.clear();
+                }
+            }
```
We agree with this proposal. Currently this is the way the check has been implemented. Resolving.

frie-soptim committed:
modified some load/gen types in pv-node6 test

- in order to have a greater coverage of possible appliance combinations

Signed-off-by: Eduard Fried <eduard.fried@soptim.de>

figueroa1395 committed:
resolve merge conflicts

Signed-off-by: Santiago Figueroa Manrique <figueroa1395@gmail.com>

figueroa1395 committed:
ruff format

Signed-off-by: Santiago Figueroa Manrique <figueroa1395@gmail.com>

figueroa1395 committed:
fix code quality

Signed-off-by: Santiago Figueroa Manrique <figueroa1395@gmail.com>

figueroa1395 committed:
format

Signed-off-by: Santiago Figueroa Manrique <figueroa1395@gmail.com>

figueroa1395 committed:
small cleanup

Signed-off-by: Santiago Figueroa Manrique <figueroa1395@gmail.com>

figueroa1395 committed:
Merge branch 'main' into feature/#185PV-Node

figueroa1395 reviewed (dismissed):
@frie-soptim @scud-soptim I'm approving the PR. Great work as always and thank you for the contribution! I'll send to merge queue, fingers crossed.

mgovers committed:
format json

Signed-off-by: Martijn Govers <Martijn.Govers@Alliander.com>

mgovers committed:
Merge branch 'main' into feature/#185PV-Node

Signed-off-by: Martijn Govers <martijn.govers@alliander.com>

mgovers committed:
fix merge conflicts

Signed-off-by: Martijn Govers <Martijn.Govers@Alliander.com>

mgovers reviewed (approved):
This PR will be merged on Monday to reduce the risk of introducing an issue and having to work over the weekend.

After that, we can create a PR to drop the feature flag and officially release the feature (I will do another round of review but as @figueroa1395 already mentioned, everything looks good).

mgovers committed:
Merge branch 'main' into feature/#185PV-Node
